### PR TITLE
Add docs and benchmark for JSON flattening parser

### PR DESCRIPTION
**Pull request by jon-wei**
Part of a set of 3 related pull requests, addressing Druid issue:
 https://github.com/druid-io/druid/issues/1839

https://github.com/metamx/java-util/pull/34   -- new JSON parser 
https://github.com/druid-io/druid-api/pull/65  -- ingestion spec modifications
https://github.com/druid-io/druid/pull/1921  -- docs and benchmark

Adds documentation and a JMH benchmark for comparing the new JsonPath-based parser with the existing parser.

Example benchmark result:

java -jar target/benchmarks.jar FlattenJSONBenchmark -wi 1 -i 250 -f 1 -v EXTRA

Result "baseline":
  19.366 ±(99.9%) 0.268 us/op [Average]
  (min, avg, max) = (17.979, 19.366, 28.522), stdev = 1.274
  CI (99.9%): [19.098, 19.634] (assumes normal distribution)

Result "flatten":
  25.113 ±(99.9%) 0.434 us/op [Average]
  (min, avg, max) = (22.311, 25.113, 37.351), stdev = 2.061
  CI (99.9%): [24.679, 25.547] (assumes normal distribution)

Run complete. Total time: 00:10:46

Benchmark                      Mode  Cnt   Score   Error  Units
FlattenJSONBenchmark.baseline  avgt  250  19.366 ± 0.268  us/op
FlattenJSONBenchmark.flatten   avgt  250  25.113 ± 0.434  us/op

### CODE REVIEW

File: benchmarks/pom.xml [hunk 1]
```diff
@@ -51,6 +51,21 @@
       <artifactId>druid-processing</artifactId>
       <version>${project.parent.version}</version>
     </dependency>
+    <dependency>
+      <groupId>com.github.wnameless</groupId>
+      <artifactId>json-flattener</artifactId>
+      <version>0.1.0</version>
+    </dependency>
+    <dependency>
+      <groupId>com.jayway.jsonpath</groupId>
```
Comment (gianm) [inline benchmarks/pom.xml]: you shouldn't have to add this, it should get pulled in by java-util now.

File: docs/content/ingestion/flatten-json.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  "name" : "metrica",
+  "fieldName" : "metrica"
+}
+]
+```
+
+Note that:
+
+* For convenience, when defining a root-level field, it is possible to define only the field name, as a string, shown with "dim2" above.
+* Enabling 'useFieldDiscovery' will only autodetect fields at the root level with a single value, not a map or list. In the example above, "dim1", "dim2", "dim3", "ignore_me", "metrica", and "foo.bar" (at the root) would be automatically detected as columns, but the fields "hello" and "world" must be explicitly defined.
```
Comment (gianm) [inline docs/content/ingestion/flatten-json.md]: this should detect fields that are Lists of simple types, so please update the docs along with the parser.

File: benchmarks/src/main/test/io/druid/benchmark/FlattenJSONBenchmarkUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    Assert.assertEquals("-7.8634787235005573E18", event.get("e3.am1[1]").toString());
+    Assert.assertEquals("8.7372945568982446E18", event.get("e3.am1[2]").toString());
+    Assert.assertEquals("3.1928124802414899E18", event.get("e3.am1[3]").toString());
+    Assert.assertEquals("-3.9806631713718011E18", event.get("e4.e4.m4").toString());
+    Assert.assertEquals("-1915243040", event.get("ae1[0].d1").toString());
+    Assert.assertEquals("-2020543641", event.get("ae1[1].d1").toString());
+    Assert.assertEquals("1414285347", event.get("ae1[2].e1.d2").toString());
+
+    /*
+    for (String key : event2.keySet()) {
```
Comment (gianm) [inline benchmarks/src/main/test/io/druid/benchmark/FlattenJSONBenchmarkUtilTest.java]: remove dead code

File: benchmarks/src/main/test/io/druid/benchmark/FlattenJSONBenchmarkUtilTest.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    StringInputRowParser nestedParser = eventGen.getNestedParser();
+
+    InputRow parsed = flatParser.parse(newEvent);
+    InputRow parsed2 = nestedParser.parse(newEvent2);
+
+    Map<String, Object> event = ((MapBasedRow) parsed).getEvent();
+    Map<String, Object> event2 = ((MapBasedRow) parsed2).getEvent();
+
+    /*
+    for (String key : event.keySet()) {
```
Comment (gianm) [inline benchmarks/src/main/test/io/druid/benchmark/FlattenJSONBenchmarkUtilTest.java]: dead code

File: benchmarks/pom.xml (hunk 1 above)
Comment (jon-wei) [inline benchmarks/pom.xml]: @gianm removed JsonPath dependency from benchmarks

File: docs/content/ingestion/flatten-json.md (hunk 2 above)
Comment (jon-wei) [inline docs/content/ingestion/flatten-json.md]: @gianm parser and docs updated

File: benchmarks/src/main/test/io/druid/benchmark/FlattenJSONBenchmarkUtilTest.java (hunk 3 above)
Comment (jon-wei) [inline benchmarks/src/main/test/io/druid/benchmark/FlattenJSONBenchmarkUtilTest.java]: @gianm removed

File: benchmarks/src/main/test/io/druid/benchmark/FlattenJSONBenchmarkUtilTest.java (hunk 4 above)
Comment (jon-wei) [inline benchmarks/src/main/test/io/druid/benchmark/FlattenJSONBenchmarkUtilTest.java]: @gianm removed

**Comment by jon-wei** [comment]
@drcrallen @gianm @himanshug 

The benchmarks were unnecessarily creating a StringInputRowParser, I've revised them to use only the JSONParser/JSONPathParser.

New benchmark numbers:

Benchmark                                    Mode  Cnt   Score   Error  Units
FlattenJSONBenchmark.baseline                avgt   50  20.971 ± 2.840  us/op
FlattenJSONBenchmark.flatten                 avgt   50  36.168 ± 7.699  us/op
FlattenJSONBenchmark.preflattenNestedParser  avgt   50  19.007 ± 3.677  us/op

baseline is old JSONParser
flatten is new JSONPathParser on nested input
preflatten is new JSONPathParser on preflattened input, same input as baseline

After discussing some profiling results with @gianm  today, I added a 4th benchmark that extracts root-level fields with JsonPath (i.e., specify a "nested" field but with a root expression). This was done to get an idea of the overhead added by JsonPath, normally the new JSONPathParser would read directly from the Jackson-provided Map for root fields.

With the same flattened input as baseline and preflatten:
FlattenJSONBenchmark.forcedRootPaths  avgt   50  30.024 ± 4.324  us/op

It looks like:
* the new parser and old parser perform similarly for preflattened input. 
* parsing my test nested input takes ~50% longer compared to a preflattened input
* JsonPath reads add significant overhead for both root fields and more deeply nested fields.

**Comment by himanshug** [comment]
Given the results, I am +1 for deprecating old JsonParser  and just using the new one.

**Comment by himanshug** [comment]
:+1:

**Comment by jon-wei** [comment]
Before merging, needs a new druid-api version that includes https://github.com/druid-io/druid-api/commit/e8c3533c8f87c5d0cbef487ea9ea6a08c54e3fbc

**Comment by jon-wei** [comment]
Updated pom.xml to use druid-api 3.14, is this good to merge?

**Comment by jon-wei** [comment]
@gianm @himanshug Do you have any more feedback for this?

**Comment by himanshug** [comment]
:+1:

**Comment by gianm** [comment]
:+1:

**Pull request merged**